src/main/java/org/apache/mesos/offer/ResourceUtils.java

```diff
@@ -203,66 +204,155 @@ public static Resource getExpectedRanges(
+        }
+        if (key != null && value != null) {
+            resReq.setVIPLabel(Label.newBuilder().setKey(key).setValue(value).build());
+        }
     }
 
-    public static Protos.Environment getEnvironment(List<Resource> resources) {
-        Protos.Environment.Builder envBuilder = Protos.Environment.newBuilder();
-        for (Resource resource : resources) {
-            String portName = DynamicPortRequirement.getPortName(resource);
-            if (portName != null) {
-                String portNumber = String.valueOf(resource.getRanges().getRange(0).getBegin());
-                envBuilder.addVariables(Protos.Environment.Variable.newBuilder()
-                        .setName(portName)
-                        .setValue(portNumber));
+    public static Resource getResourceAddLabelUnique(Resource resource, Label label) {
+        Resource.ReservationInfo.Builder reservationBuilder = Resource.ReservationInfo
+                .newBuilder(resource.getReservation());
+        Resource.Builder resourceBuilder = Resource.newBuilder(resource);
+
+        List<Label> labelList = resource.getReservation().getLabels().getLabelsList();
+
+        Labels.Builder labelsBuilder = Labels.newBuilder();
+        for (Label labelIn : labelList) {
+            if (!(labelIn.getKey()).equals(label.getKey())) {
+                labelsBuilder.addLabels(labelIn);
             }
         }
+        labelsBuilder.addLabels(label);
 
-        return envBuilder.build();
+        reservationBuilder.setLabels(labelsBuilder.build());
+        resourceBuilder.setReservation(reservationBuilder.build());
+        return resourceBuilder.build();
     }
 
-    public static Protos.Environment updateEnvironment(Protos.Environment env, List<Resource> resources) {
-        Protos.Environment resEnv = ResourceUtils.getEnvironment(resources);
-        return Protos.Environment.newBuilder(env)
-                .addAllVariables(resEnv.getVariablesList())
-                .build();
+    public static TaskInfo.Builder setVIPDiscovery(TaskInfo.Builder builder, String execName,
+                                                   List<Resource> resourceList) {
+        TaskInfo.Builder taskBuilder = builder;
+
+        for (Resource resource : resourceList) {
+            ResourceRequirement resReq = new ResourceRequirement(resource);
+                taskBuilder = setVIPDiscovery(taskBuilder, execName, resReq);
+        }
+        return taskBuilder;
+    }
+
+    public static TaskInfo.Builder setVIPDiscovery2(TaskInfo.Builder builder, String execName,
+                                                    List<ResourceRequirement> resReqList) {
+        TaskInfo.Builder taskBuilder = builder;
+
+        for (ResourceRequirement resReq : resReqList) {
+            taskBuilder = setVIPDiscovery(taskBuilder, execName, resReq);
+        }
+        return taskBuilder;
+    }
+
+    /* what if we have multiple ports, or multiple VIPs, we need to work on here !!!! */ //FIX Later
```

Comment:
I don't understand this variable name.  `execName`?  Is it from Executor?  There are no executors in this code.